File: backend/app/schemas/social.py
from typing import List, Optional, TypeVar, Generic

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel

T = TypeVar("T")


class ResponseEnvelope(BaseModel, Generic[T]):
    code: int = 200
    msg: str = "success"
    data: T


class SocialBaseModel(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel, populate_by_name=True)


class CommentCreate(SocialBaseModel):
    content: str = Field(..., min_length=1, max_length=500)


class DynamicUpdate(SocialBaseModel):
    content: str = Field(..., min_length=1, max_length=2000)


class LikeResponse(SocialBaseModel):
    is_liked: bool
    like_count: int


class CommentResponse(SocialBaseModel):
    id: str
    user_id: str
    username: str
    avatar: str
    dynamic_id: str
    content: str
    created_at: str
    like_count: int = 0
    is_liked: bool = False
    is_owner: bool


class LikedUserBrief(SocialBaseModel):
    user_id: str
    username: str
    avatar: str


class CommentBrief(SocialBaseModel):
    user_id: str
    username: str
    avatar: str
    content: str
    created_at: str


class DynamicResponse(SocialBaseModel):
    id: str
    user_id: str
    cat_id: Optional[str] = None
    username: str
    cat_name: Optional[str] = None
    avatar: str
    content: str
    images: List[str] = Field(default_factory=list)
    like_count: int
    comment_count: int
    favorite_count: int = 0
    created_at: str
    is_recommended: bool = False
    recommended_at: Optional[str] = None
    is_liked: bool = False
    is_favorited: bool = False
    is_following: bool = False
    is_owner: bool = False
    comments: List[CommentResponse] = Field(default_factory=list)
    recent_likers: List[LikedUserBrief] = Field(default_factory=list)
    latest_comment: Optional[CommentBrief] = None


class DynamicsListResponse(SocialBaseModel):
    list: List[DynamicResponse]
    total: int = 0
    page: int = 1
    page_size: int = 10
    cursor: Optional[str] = None
    limit: int
    has_more: bool = False

class FollowResponse(SocialBaseModel):
    is_following: bool
    follower_count: int

class FavoriteResponse(SocialBaseModel):
    is_favorited: bool
    favorite_count: int

class HotTopic(SocialBaseModel):
    topic: str
    count: int
    is_default: bool = False

class HotTopicsResponse(SocialBaseModel):
    list: List[HotTopic]

class DeleteResponse(SocialBaseModel):
    success: bool

class FollowerResponse(SocialBaseModel):
    id: str
    user_id: str
    username: str
    nickname: str
    avatar: str
    last_online_at: str
    is_online: bool
    is_following: bool

class FollowersListResponse(SocialBaseModel):
    list: List[FollowerResponse]
    total: int
    page: int
    page_size: int


class LikerResponse(SocialBaseModel):
    user_id: str
    username: str
    avatar: str


class LikerListResponse(SocialBaseModel):
    list: List[LikerResponse]
    total: int
    page: int
    page_size: int

File: frontpoint/src/types/social.ts
// 社交功能相关类型定义

// 发布动态请求
export interface PublishDynamicRequest {
  catId: string;
  content: string;
  images?: File[]; // 图片文件数组（multipart/form-data）
}

// 动态响应
export interface DynamicResponse {
  id: string;
  userId: string;
  catId?: string;
  username: string;
  catName?: string;
  avatar: string;
  content: string;
  images?: string[];
  likeCount: number;
  commentCount: number;
  favoriteCount?: number;
  createdAt: string;
  isLiked?: boolean;
  isFavorited?: boolean;
  isFollowing?: boolean;
  isOwner: boolean; // 是否是自己的动态
  comments?: CommentResponse[];
}

// 动态列表请求参数
export interface DynamicsListRequest {
  catId?: string; // 可选，指定猫咪的动态
  userId?: string; // 可选，指定用户的动态
  page: number;
  pageSize: number;
}

// 动态列表响应
export interface DynamicsListResponse {
  list: DynamicResponse[];
  total: number;
  page: number;
  pageSize: number;
}

// 评论请求
export interface CommentRequest {
  dynamicId: string;
  content: string;
}

// 评论响应
export interface CommentResponse {
  id: string;
  userId: string;
  username: string;
  avatar: string;
  dynamicId: string;
  content: string;
  createdAt: string;
  likeCount?: number;
  isLiked?: boolean;
  isOwner: boolean; // 是否是自己的评论
}

// 点赞请求
export interface LikeRequest {
  dynamicId: string;
}

// 点赞响应
export interface LikeResponse {
  isLiked: boolean;
  likeCount: number;
}

export interface FavoriteResponse {
  isFavorited: boolean;
  favoriteCount: number;
}

export interface FollowResponse {
  isFollowing: boolean;
  followerCount: number;
}

// 我的动态列表请求参数
export interface MyDynamicsListRequest {
  catId?: string;
  page: number;
  pageSize: number;
}

// 我的动态列表响应
export interface MyDynamicsListResponse {
  list: DynamicResponse[];
  total: number;
  page: number;
  pageSize: number;
}

export interface HotTopicItem {
  topic: string;
  count: number;
}

export interface HotTopicsResponse {
  list: HotTopicItem[];
}

// Backward-compatible aliases for older API/composable names.
export type SocialDynamic = DynamicResponse;
export type SocialDynamicListResponse = DynamicsListResponse;
export type SocialDynamicResponse = DynamicResponse;
export type DynamicDetailResponse = DynamicResponse;
export type LikeDynamicResponse = LikeResponse;